Create Reactions View

kennethnrk opened the pull request:
# Create Reactions View
- Created a new search operator `has:reaction`
- Added search suggestions for the new operator
- Created a new view in the left sidebar for `has:reaction sender:me`
- Updated `web/src/filter.ts` and `web/src/message_parser.ts` to deal with cached messages
- Fixed `message_reactions` undefined by cleaning reactions earlier in the flow.

## Differences from previous plans
- Used `zulip-icon-smile` instead of https://feathericons.com/?query=thumbs-up to maintain uniformity with other reaction buttons.

Fixes: [Issue 27328](https://github.com/zulip/zulip/issues/27328)

**Screenshots and screen captures:**
### Tooltip and custom heading
![image](https://github.com/zulip/zulip/assets/29176437/04802254-8d7e-4a3a-b268-f6e7e65a7a88)
### Help Menu
![image](https://github.com/zulip/zulip/assets/29176437/9105479d-efd8-4c94-aaaa-02a1b374b883)

### Empty narrow banner
![image](https://github.com/zulip/zulip/assets/29176437/077e07a6-8a9b-4bbe-a3ab-b7953527001a)

### Collapsed Sidebar
![image](https://github.com/zulip/zulip/assets/29176437/b9dd8ca1-2123-4596-a21a-e809b1e389ca)



<details>
<summary>Self-review checklist</summary>

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/narrow.js:
```diff
@@ -209,6 +209,10 @@ export function activate(raw_terms, opts) {
         }
+        if (filter.has_operand("has","reaction") && !(filter.has_operator("sender")) ) {
+            console.log("Pushing sender to filter")
+            filter._terms.push({operator:"sender", operand:people.my_current_email(), negated:false})
```
I don't understand this change; I think the idea was we'd just have a link to the `sender:me has:reaction` search query, not that we'd make it impossible to search for just `has:reaction`, as this proposes.

timabbott replied:
It looks like you remove this in a future commit in this PR. In general, you should clean that sort of thing up before asking for review, as it can be confusing and waste reviewer time that would be better spent on changes that you think are correct.

Check out the [Zulip commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages) and today's community event for more details.

timabbott commented on web/src/message_parser.ts:
```diff
@@ -23,3 +23,7 @@ export function message_has_image(message: Message): boolean {
 }
+
+export function message_has_reaction(message: Message): boolean {
+    return message?.message_reactions?.length > 0;
```
This seems wrong given the types. If you accept a `Message` object, we don't need to consider it being `undefined`. Likely you have a bug in the caller that you're working around.

Also the red symbol just below this comment indicates you deleted the newline at the end of this file; you should add one since that's always preferred, and maybe configure your editor to always generate them.

timabbott commented on zerver/migrations/0501_merge_20240228_1140.py:
```diff
@@ -0,0 +1,13 @@
+        ("zerver", "0500_realm_zulip_update_announcements_stream"),
+    ]
+
+    operations = []
```
You should be renumbering migrations when you rebase, not creating these merge migrations as Django's tooling suggests. See https://zulip.readthedocs.io/en/latest/subsystems/schema-migrations.html for important documentation on this topic.

timabbott commented on zerver/migrations/0497_archivedmessage_has_reaction_message_has_reaction.py:
```diff
@@ -0,0 +1,23 @@
+            name="has_reaction",
+            field=models.BooleanField(db_index=True, default=False),
+        ),
+    ]
```
You have one more major bug in this code, which is that you need to write a backfill migration that 

Read `zerver/migrations/` to understand some examples; `ls zerver/migrations/*backfill*` should give several useful examples.

timabbott commented:
So I posted some feedback on how maintaining a `has_reaction` database field requires writing an expensive backfill migration. Thinking about this more, I also wonder if that's necessary? I think you should be able to just write a database query in the `narrow.py` code using SQLAlchemy to join against `zerver_reaction` and filter to messages that contain a row, and that database query might be well-indexed already, since `Reaction`.

So you might want to drop adding a new field from this PR and see if you can get that to work well.

kennethnrk commented on zerver/lib/narrow.py:
Majority of the changes are in this file. 

I needed to specify the table name along with the column name in order to avoid ambiguity between columns of `zerver_usermessages` and `zerver_reaction`.

Added the `need_user_message` param to check if `zerver_usermessage.flags` should be included in the group by query or not.

kennethnrk commented:
I have reverted the previous commit that was modifying the base query and I have implemented the join using a subquery instead.

I had added the `need_user_message` param to the `add_narrow_condition`. I tried figure out a workaround to this, but I think it should be safe because `need_user_message` was already defined in the functions calling `add_narrow_condition`. If anyone has any workaround in terms of how to implement the `ORDER BY` clause without passing the `need_user_message` param, please suggest.

Thanks.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -333,8 +335,24 @@ def add_term(self, query: Select, term: Dict[str, Any]) -> Select:
+                return query.join(
+                    subquery,
+                    self.msg_id_column == literal_column("zerver_reaction_subtable.reaction_message_id", Integer),
+                ).group_by(self.msg_id_column)
```
We should do a fancier check that just does an `exists` check, since that's all we care about here. Might make this a bit more efficient.

kennethnrk commented:
amended commit [Distinct: Use distinct instead of group by](https://github.com/zulip/zulip/pull/29109/commits/8cfbc2e663c85afe0561c79aeebc6b7892378667) to pass linting

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

You will also need to make sure that your PR passes tests.

timabbott commented on web/src/message_helper.js:
```diff
@@ -25,6 +25,7 @@ export function process_new_message(message) {
 
     message_store.set_message_booleans(message);
     message.sent_by_me = people.is_current_user(message.sender_email);
+    message.message_reactions = message.reactions || [];
```
This seems very suspect to me; can you talk about what made you think this change should be made?

`reactions.js` has some logic involving `clean_reactions` that may be related.

timabbott replied:
Yeah, that seems OK, with a comment, as we'd just be doing that normal processing a little early.

timabbott commented:
OK this is starting to look close in terms of the final state of the branch. Can you clean up the commit history to have a series of clear, clean commits with readable commit messages, as per our contributing guidelines? Reading a few recently merged PRs with several commits will likely provide helpful examples.

Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for more details.

timabbott commented on web/src/message_helper.js:
```diff
@@ -23,9 +24,17 @@ export function process_new_message(message) {
 
+    // Cleaning message_reactions during initial message processing
+    // to avoid message_reactios being undefined during rendering
+    message.message_reactions = reactions.get_message_reactions(message);
```
You have a typo in the comment -- please proofread your comments.

In any case, @evykassirer @amanagr am I correct that this change should mean we can/should delete the similar calls to `get_message_reactions` in `message_list_view.js`? 
```
            const message_reactions = reactions.get_message_reactions(message_container.msg);           
            message_container.msg.message_reactions = message_reactions;                                
```

It seems like if we can assume that this happens before messages are inserted into the message store, we should be able to not call it a second time in other code paths...

This commit might be worth extracting to its own miniature pull request that would contain those changes and also changes to the `Message` type to simplify the main `export type Message`.

amanagr replied:
Seems correct given that `server_events_dispatch` also updates it.

kennethnrk replied:
I have split the PR into [#29109 ( This PR )](https://github.com/zulip/zulip/pull/29109) and [#29280 Clean meassge reactions during initial processing.](https://github.com/zulip/zulip/pull/29280) and have removed the redundant calls to the `get_message_reactions`.

I have attempted to clean up and simplify `export type Message`.
Please give feedback on how more relevant changes that can be made to simplify the type Message.

alya commented:
> Used zulip-icon-smile instead of https://feathericons.com/?query=thumbs-up

Can you explain why you decided to make this change? If there was a discussion on CZO, please link to it from the PR description.

alya commented:
Please add a screenshot of how this view is shown when Views are collapsed.

kennethnrk commented:
> > Used zulip-icon-smile instead of https://feathericons.com/?query=thumbs-up
> 
> Can you explain why you decided to make this change? If there was a discussion on CZO, please link to it from the PR description.

No particular reason other than it looking more uniform with the other emoji buttons. Should I have started a discussion on CZO regarding this?

alya commented:
Not necessarily, but the PR should *explain* the differences from previous plans, not just list them. :)

alya commented:
> We can try putting it last in the three-dot menu -- I'm not sure what's best.

Your screenshot doesn't match this -- this should be inside the menu, not visible when views are collapsed.

kennethnrk commented:
> > We can try putting it last in the three-dot menu
> 
> Your screenshot doesn't match this -- this should be inside the menu, not visible when views are collapsed.



> > We can try putting it last in the three-dot menu
> 
> Your screenshot doesn't match this -- this should be inside the menu, not visible when views are collapsed.

Ok, I've made the change.

alya commented:
Cool, thanks. 

Live update doesn't seem to be working here. If I send a message and react to it in another tab, it does not immediately show up in this view.

alya commented:
The view should also have a custom heading, like other views in the left sidebar:

<img width="242" alt="Screenshot 2024-03-14 at 2 29 39 PM" src="https://github.com/zulip/zulip/assets/2090066/667a16e1-3178-4f9b-a8f8-ff292f3265e4">

alya commented:
We should probably also add a tooltip to the left sidebar item explaining what exactly this is. Maybe: "Your messages that have received reactions."

alya commented:
We can try it out, but I'm not sure at this point whether we'll want to display this view in the left sidebar, or put it in a `...` menu even when the Views section is expanded.

kennethnrk commented:
> We should probably also add a tooltip to the left sidebar item explaining what exactly this is. Maybe: "Your messages that have received reactions."

I have added the custom heading and tooltips.

kennethnrk commented:
> Live update doesn't seem to be working here. If I send a message and react to it in another tab, it does not immediately show up in this view.

Can you please guide me as to which function needs to called in order to make this work.

I have tried calling `message_live_update.rerender_messages_view();`  when a new reaction is added, but it's not working.

alya commented:
> Can you please guide me as to which function needs to called in order to make this work.

You can ask in the #frontend stream on chat.zulip.org.

kennethnrk commented:
I have split the the [backend ](https://github.com/zulip/zulip/pull/29560) and the frontend. 

I have added some comments to [clean message reactions during preprocessing](https://github.com/zulip/zulip/pull/29280) on how I think we should move forward with it.

alya commented:
We decided on CZO not to implement live-updated. When the rest of the feedback above has been addressed, please update the screenshots in the PR description and post a comment here to say that it's ready.

Also, could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

kennethnrk commented:
> We decided on CZO not to implement live-updated. When the rest of the feedback above has been addressed, please update the screenshots in the PR description and post a comment here to say that it's ready.
> 
> Also, could you please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages)?

Added the relevant screenshots and fixed the commits.

alya commented:
Let's change the help menu documentation description to "Your messages with reactions." I think it's clear enough, and should fit better.

kennethnrk commented:
> Let's change the help menu documentation description to "Your messages with reactions." I think it's clear enough, and should fit better.

I'm not sure that would be accurate, the `has:reaction` operator will return everyone's messages with reactions, not just "Your messages with reaction". Instead I could change it to "Messages with reactions"?

kennethnrk commented:
>  Instead I could change it to "Messages with reactions"?
We could also just go with "Messages containing reactions" to maintain uniformity with other has operators?

alya commented:
Oh, good point. "Messages with reactions" is good, then.

alya commented:
"Containing" won't really make sense, since the reaction isn't contained in the message the same way that mentions, etc. are.

alya commented:
I think this doesn't work when you have no messages with reactions? I just got this error when clicking on "Reactions" in the left sidebar.

![CleanShot 2024-04-12 at 08 56 25@2x](https://github.com/zulip/zulip/assets/2090066/19822e26-c7fb-427f-a2bc-2fa5deb7a47a)

kennethnrk commented:
> I think this doesn't work when you have no messages with reactions? 


This PR depends on [#29280](https://github.com/zulip/zulip/pull/29280) it doesn't work without it.

alya commented:
Ok, then that status should be made clear in the PR description, to avoid wasting reviewer time.

kennethnrk commented:
> Ok, then that status should be made clear in the PR description above, to avoid wasting reviewer time.

I'm sorry, the comment got lost among the other comments above. I will mention it in the description also.

kennethnrk commented:
@alya The backend as well as the message reactions prerequisite PRs have been merged. I have rebased the branch. Please continue your review, sorry for the hiccup.

alya commented:
Thanks for the update!

alya commented:
We need a custom placeholder when there are no messages to show. This doesn't work for a view that can be accessed from the left sidebar:

![CleanShot 2024-04-15 at 15 37 15@2x](https://github.com/zulip/zulip/assets/2090066/52ccb6c4-3f1b-4141-80fd-5af027f1e85e)

We can follow the same pattern as for @-mentions:

> You have no messages with reactions yet!
> Learn more about reactions [here](/help/emoji-reactions).

alya commented:
The "End of results" notice should be removed in general.

![CleanShot 2024-04-15 at 15 41 56@2x](https://github.com/zulip/zulip/assets/2090066/bd2df16e-ba3e-4d1b-8851-08d51e217e64)

alya commented:
In the menu when views are collapsed, there is an incorrect counter shown.

![CleanShot 2024-04-15 at 15 44 22@2x](https://github.com/zulip/zulip/assets/2090066/a2b41989-9e98-407b-bfc8-70cb2c88f372)

alya commented:
It looks like the left sidebar link tooltip doesn't have an appropriate delay. The delay should match other Views.

alya commented:
Drafts and Scheduled messages should be next to each other in the menu. We can try putting Reactions first there, which will be more consistent with the order we've got in the expanded list of views.

![CleanShot 2024-04-15 at 15 46 53@2x](https://github.com/zulip/zulip/assets/2090066/ae80847a-3ce9-46b0-8833-ea8573b3e117)

alya commented:
The tooltip in the menu is broken.

![Uploading CleanShot 2024-04-15 at 15.48.59@2x.png…]()

alya commented:
It looks to me like the menu when views are collapsed was not manually tested, or was tested very lightly. Please be mindful to thoroughly test your own work before asking others for review.

kennethnrk commented:
> It looks to me like the menu when views are collapsed was not manually tested, or was tested very lightly. Please be mindful to thoroughly test your own work before asking others for review.

I have made the required changes, going forward i will be more diligent with my testing.

alya commented:
Thanks! Please add screenshots to the PR description to demonstrate that the issues have been fixed.

kennethnrk commented:
> Thanks! Please add screenshots to the PR description to demonstrate that the feedback has been addressed.

Added the screenshots to the description.

kennethnrk commented:
Since #30227 has been merged, I think this PR is ready to merge.

timabbott commented:
* I think we might want the description to be more like "Reactions to your messages"; probably worth discussing in the design thread for the feature.

timabbott commented on help/search-for-messages.md:
```diff
@@ -96,6 +96,7 @@ Zulip offers the following filters based on the location of the message.
 * `has:attachment`: Search messages that contain an [uploaded
   file](/help/share-and-upload-files).
 * `has:image`: Search messages that contain uploaded or linked images or videos.
+* `has:reaction`: Search messages with [reactions](/help/emoji-reactions).
```
I think probably spelling out "emoji reactions" in the description here is appropriate.

timabbott commented on web/src/filter.ts:
```diff
@@ -862,6 +869,10 @@ export class Filter {
 
+        if (_.isEqual(term_types, ["sender", "search", "has-reaction"])) {
+            return "/#narrow/has/reaction/sender/me";
+        }
```
Can you explain this logic? It seems like it ignores the actual operand for the current view, so if we were looking at reactions sent by some other user, this would switch it to us?

kennethnrk replied:
Yes, I have changed it to also check `filter.operands("sender")[0] === people.my_current_email()` wherever required.

timabbott commented on web/src/narrow_banner.ts:
```diff
@@ -160,6 +162,23 @@ function pick_empty_narrow_banner(): NarrowBannerData {
 
+        if (_.isEqual(current_terms_types, ["sender", "has-reaction"])) {
+            return {
+                title: $t({defaultMessage: "You have no messages with reactions yet!"}),
```
This needs something more precise, like "None of your messages have emoji reactions yet." 

Absent clarity, someone could read "your messages" as the ones they've sent or the ones they've received.

timabbott commented on web/templates/popovers/left_sidebar/left_sidebar_condensed_views_popover.hbs:
```diff
@@ -1,5 +1,14 @@
 {{! Contents of the condensed nav vdots popup }}
 <ul class="nav nav-list condensed-views-popover-menu">
+    <li class="condensed-views-popover-menu-drafts">
```
This class name seems incorrect, it should say `condensed-views-popover-menu-reactions`

Be really careful whenever you copy/paste to actually check that you changed everything.

@alya FYI, this is ![image](https://github.com/zulip/zulip/assets/2746074/e37905a9-45e7-40bf-8d5a-1c994cbe94ca), just making sure you saw that state as a thing to test.

alya replied:
Yes, lots of feedback on that above.

timabbott commented:
This lgtm aside from the details noted above. Probably the string details should be discussed and finalized in the #design or #feedback thread for the feature.

alya commented:
I think "Reactions" as the main label, and "Reactions to your messages" as the tooltip should be clear enough.

alya commented:
Oh, and we should put "Reactions to your messages." as the view description in the top navbar.

kennethnrk commented:
> Oh, and we should put "Reactions to your messages." as the view description in the top navbar.

Should I make it "Emoji reactions to your messages."? Do we need a #design thread to discuss this?

kennethnrk commented:
I have made the string changes (used "emoji reactions" instead of "reactions"), and changes to the `generate_redirect_url` logic and any similar functions. I have updated the screenshots in the descriptions.

timabbott commented:
Looks great, merged, thanks for all the work on this @kennethnrk! It turned out to be quite a journey :)
